### DISCUSSION TITLE
Sample code for calling launcher menu item from CLI

### DISCUSSION THREAD

**Discussion by mitchcapper** [wezterm/wezterm#3836] [Show and tell]
The launcher menu is an array of SpawnCommand ( https://wezfurlong.org/wezterm/config/lua/SpawnCommand.html ) elements which offer through configuration for new tabs/windows.   There isn't a direct connection to call a launch menu item from the CLI (that I am aware of) but it is possible to work around it pretty easily.  We can tap the gui-startup event.  We can't modify the parameter passed to gui-startup (a SpawnCommand representing the users initial CLI call) but we can just manually handle the spawning ourselves.  If we spawn a window during the event handler the default action is suppressed.

This allows for easily creating say windows shortcuts to specific menu items as well.

How you want to handle the CLI invoking can be adjusted but the simple example below allows overloading of the command itself with any label from the launch menu.  For example calling: `wezterm-gui start -- BashCon`   .  If you run into an issue just run the command from an existing shell with the base wezterm rather than wezterm-gui executable (ie `wezterm start -- BashCon`).  If the issue still isn't obvious uncomment the log line to see what it is finding vs the calling arg.  It would be trivial to do things like tack additional command line args onto the pre-defined (if any) launch args for the menu item,  or to say require a verb prior to the menu item name so you have to to do `wezterm start -- launch_item BashCon`.

Here is a full sample config (should work on releases of wezterm from the last few months).
```lua
local wezterm = require 'wezterm'
local act = wezterm.action
local mux = wezterm.mux
local config = wezterm.config_builder()
wezterm.on('gui-startup', function(spawnCmd)
  if spawnCmd then
  	if #(spawnCmd.args) > 0 then
  		local findName = string.lower(spawnCmd.args[1])
  		for _, menuItem in ipairs(config.launch_menu) do
			-- wezterm.log_info("Comparing launch first arg of: " .. findName .. " to: " .. string.lower(menuItem.label))
			if string.lower(menuItem.label) == findName then
				mux.spawn_window(menuItem)
			end
		end
  	end
  end
end)

config.launch_menu = {
	{
		label = 'VS Powershell',
		args = {'v:/Users/VirtM/AppData/Local/Microsoft/WindowsApps/pwsh.exe', '-NoExit', '-Command', '&{Import-Module "v:/Program Files/Microsoft Visual Studio/2022/Preview/Common7/Tools/Microsoft.VisualStudio.DevShell.dll"; Enter-VsDevShell -SkipAutomaticLocation -DevCmdArguments "-arch=x64 -host_arch=x64"}'},
		cwd = 'v:/checkout/'
	},
	{
		label = 'MSYS VS Powershell',
		args = {'v:/Users/VirtM/AppData/Local/Microsoft/WindowsApps/pwsh.exe', '-NoExit', '-Command', 'v:/checkout/WIN64LinuxBuild/vs_msys_shell_launch.ps1'},
		cwd = 'v:/checkout/'

	},
	{
		label = 'PowerShell Classic 5',
		args = {'powershell.exe', '-NoLogo'}
	},
	{
		label = 'BASHCon',
		args = {'v:/bin/bash.exe'},
		set_environment_variables = {
			HOME = "v:/users/VirtM",
			SHELL = "v:/bin/bash.exe"
		}
	},
}
return config
```

**Comment by DeadlySquad13** [comment]
Big thanks, I was searching through the docs for this feature for a while until found your tip by accident!

One small detail I want to point out that I've stumbled upon:

> If we spawn a window during the event handler the default action is suppressed

Be careful with that thing. In my case wezterm did launch the default action, so two windows were spawned: one corresponding to launch menu and another - just a process.

In [docs](https://wezterm.org/config/lua/gui-events/gui-startup.html?h=gui+startup) it's written quite ambiguously: 

> In earlier releases if you implemented this event, you would essentially prevent wezterm start -- something from spawning something.

So I almost gave up - it seemed to me that it was deprecated. In reality it was just my second callback on this listener that was breaking this intricate detail:

```lua
wezterm.on('gui-startup', function(cmd)
  local tab, pane, window = mux.spawn_window(cmd or {})
  window:gui_window():maximize()
end)
```

It was spawning window so I had this behaviour.

Combining into one callback and removing extra window_spawn in the branch of
a `launch_menu` solved my issue.

Final code:

```lua

local function start_fullscreen(cmd)
  local _, _, window = mux.spawn_window(cmd or {})

  -- Fullscreen on startup.
  -- If you leave fullscreen, stay maximized. Removed because of komorebi.
  -- window:gui_window():maximize()
  window:gui_window():toggle_fullscreen()
end


-- Example:
-- ```bash
-- wezterm start -- PowerShell # Follow default processing of args: run PowerShell in new window.
-- wezterm start -- launch_menu PowerShell # Spawn new window from launch_menu labeled "PowerShell"
-- ```
--
-- [Source]: <https://github.com/wezterm/wezterm/discussions/3836> 'Sample Code for Calling Launcher Menu Item from CLI · Wezterm/Wezterm · Discussion \#3836'
local function transform_cmd(cmd)
  if not cmd or #(cmd.args) < 2 then
    return cmd
  end

  local action_type = string.lower(cmd.args[1])

  if action_type ~= "launch_menu" then
    return cmd
  end

  local findName = string.lower(cmd.args[2])
  for _, menuItem in ipairs(config.launch_menu) do
    if string.lower(menuItem.label) == findName then
      return menuItem
    end
  end

  return cmd
end

wezterm.on("gui-startup", function(cmd)
  return start_fullscreen(transform_cmd(cmd))
end)
```
